File: F# Workspace/07_Stacks/07_Stacks/stack_menu.fs
(*
 * stack_menu.fs
 * By Antonio F. Huertas
 * This program manages a menu of common stack operations.
 *)

open FunctionalCollections    // for the Stack module functions
open System                   // for the .NET Console class methods

// Process the push element menu option.
let process_push stk =
    Console.Write "Enter a character: "
    let elem = Console.ReadLine () |> char
    Console.WriteLine ("Pushing {0} into the stack.", elem)
    Stack.push elem stk

// Process the pop element menu option.
let process_pop stk =
    try
        let elem = Stack.peek stk |> char
        Console.WriteLine ("Popping {0} from the stack.", elem)
        Stack.pop stk
    with
        EmptyStack msg -> Console.WriteLine ("Cannot pop: {0}", msg)
                          stk

// Process the view contents menu option.
let process_view stk =
    if Stack.isEmpty stk then
        Console.WriteLine "Cannot show: stack is empty."
    else
        Console.Write "Contents of the stack: "
        Stack.iter (printf "%c ") stk
        Console.WriteLine ()

// Shows the menu options.
let show_menu () =
    Console.WriteLine "\nStack Menu\n"
    Console.WriteLine "1. Push an element"
    Console.WriteLine "2. Pop an element"
    Console.WriteLine "3. View the contents"
    Console.WriteLine "4. Exit"

// Manages the menu selection on the stack.
let rec process_menu stk =
    show_menu ()
    Console.Write "Enter selection (1-4): "
    match (Console.ReadLine () |> int) with
    | 1 -> stk |> process_push |> process_menu
    | 2 -> stk |> process_pop |> process_menu
    | 3 -> process_view stk
           process_menu stk
    | 4 -> Console.WriteLine "Thanks for using this program!"
    | _ -> Console.WriteLine "Error! Selection must be between 1 and 4."
           process_menu stk

// Serves as the entry point of the program.
let main () =
    process_menu Stack.empty
    Console.ReadKey () |> ignore

// Calls the main function.
main()

File: C_Projects/06_Stacks/stack_menu.c
/*
 * stack_menu.c
 * By Antonio F. Huertas
 * This program manages a menu of common stack operations.
 */

#include "stack.h"   // for Stack and its functions
#include <stdio.h>   // for printf, scanf, getchar, puts
#include <stddef.h>  // for NULL definition
#include <stdlib.h>  // for malloc, free

// Declares functions using prototypes.
void show_menu(void);
int read_selection(void);
void process_selection(int sel, Stack *stk);
void process_push(Stack *stk);
void process_pop(Stack *stk);
void process_view(const Stack *stk);
void show_character(const void *elem);
void destroy_character(const void *elem);

// Starts the execution of the program.
int main(void) {
    Stack *stk = create_stack();
    puts("A stack has been created.");

    int sel = 0;
    do {
        show_menu();
        sel = read_selection();
        process_selection(sel, stk);
    } while (sel != 0);

    iterate_stack(stk, &destroy_character);
    destroy_stack(stk);
    stk = NULL;
    puts("The stack has been deallocated.");
    return 0;
}

// Shows the menu options.
void show_menu(void) {
    puts("\nStack Menu\n");
    puts("1. Push an element");
    puts("2. Pop an element");
    puts("3. View the contents");
    puts("0. Exit");
}

// Returns the menu option selected.
int read_selection(void) {
    int sel;
    printf("Enter your selection: ");
    scanf("%d", &sel);
    while (getchar() != '\n');  // flush keyboard buffer
    return sel;
}

// Processes the given menu selection and updates the stack.
void process_selection(int sel, Stack *stk) {
    switch (sel) {
        case 1: process_push(stk); break;
        case 2: process_pop(stk); break;
        case 3: process_view(stk); break;
        case 0: puts("Thanks for using this program."); break;
        default: puts("Error! Selection must be between 0 and 3.");
    }
}

// Process the push element menu option.
void process_push(Stack *stk) {
    char *elem = malloc(sizeof(char));
    printf("Enter a character: ");
    scanf("%c", elem);
    push(stk, elem);
    printf("%c has been pushed into the stack.\n", *elem);
}

// Process the pop element menu option.
void process_pop(Stack *stk) {
    char *elem = pop(stk);
    if (elem == NULL)
        puts("Could not pop.  Stack is empty.");
    else {
        printf("%c has been popped from the stack.\n", *elem);
        free(elem);
    }
}

// Process the view contents menu option.
void process_view(const Stack *stk) {
    if (is_empty_stack(stk))
        puts("Cannot show.  Stack is empty.");
    else {
        printf("These are the contents of the stack: ");
        iterate_stack(stk, &show_character);
        printf("\n");
    }
}

// Displays a character.
void show_character(const void *elem) {
    printf("%c ", *(char *)elem);
}

// Deallocates the memory of a character.
void destroy_character(const void *elem) {
    free((char *)elem);
}
